lib/classes/ConfigSchemaHandler.js

```diff
@@ -0,0 +1,143 @@
+
+const Ajv = require('ajv');
+const _ = require('lodash');
+const schema = require('../configSchema');
+const errorMessageBuilder = require('./ConfigSchemaHandler/errorMessageBuilder');
+
+const FUNCTION_NAME_PATTERN = '^[a-zA-Z0-9-_]+$';
+const ERROR_PREFIX = 'Configuration error:';
+
+class ConfigSchemaHandler {
+  constructor(serverless) {
+    this.serverless = serverless;
+    this.schema = _.cloneDeep(schema);
+
+    deepFreeze(this.schema.properties.service);
+    deepFreeze(this.schema.properties.plugins);
+    deepFreeze(this.schema.properties.package);
+    Object.freeze(this.schema.properties.layers);
+    Object.freeze(this.schema.properties.resources);
+  }
+
+  validateConfig(userConfig) {
+    const ajv = new Ajv();
+    const validate = ajv.compile(this.schema);
+
+    // console.log('userConfig', require('util').inspect(userConfig, false, null, true));
+    // console.log('=====');
+    // console.log('schema', require('util').inspect(this.schema, false, null, true));
+
+    // In case provider name is not 'aws' and no other provider schema was introduced,
+    // then this means that no other provider plugin was loaded. In this case
+    // we need to throw an error.
+    if (
+      this.serverless.service.provider.name !== 'aws' &&
+      this.schema.properties.provider.properties.name === undefined
+    ) {
+      this.handleErrorMessages([
+        `Unsupported provider name '${this.serverless.service.provider.name}'`,
+      ]);
+    }
+
+    validate(userConfig);
+    if (validate.errors) {
+      // console.log('ajv errors', require('util').inspect(validate.errors, false, null, true));
+      const messages = errorMessageBuilder.buildErrorMessages(validate.errors, userConfig);
+      this.handleErrorMessages(messages);
+    }
+  }
+
+  handleErrorMessages(messages) {
+    for (const message of messages) {
+      switch (this.serverless.service.configValidationMode) {
+        case 'error':
+          throw new this.serverless.classes.Error(`${ERROR_PREFIX} ${messages.join('; ')}`);
+        case 'warn':
+          this.serverless.cli.log(`${ERROR_PREFIX} ${message}`, 'Serverless', { color: 'red' });
+          break;
+        default:
+          throw new this.serverless.classes.Error(
+            `${ERROR_PREFIX} .configValidationMode should be equal to one of the allowed values 'error' or 'warn'`
+          );
+      }
+    }
+  }
+
+  defineTopLevelProperty(name, subSchema) {
+    this.schema.properties[name] = subSchema;
+  }
+
+  defineProvider(name, options = {}) {
+    // TODO: Remove this 'if' statement and place this piece of schema inside
+    // initial schema. This condition was added because some other tests
+    // were failing, see https://travis-ci.org/github/serverless/serverless/jobs/663153831
+    if (!this.schema.properties.provider.properties.name) {
+      this.schema.properties.provider.properties.name = { enum: [] };
+    }
+
+    this.schema.properties.provider.properties.name.enum.push(name);
```

Comment:
As I think of it, we should support only single _provider_. Therefore the logic should be more as:

```javascript
if (providerAlreadyDefined) {
  logWarning("Schema for provider was already defined"); // crash in v2
  return;
}

this.schema.properties.provider.properties.name = { enum: [name] }
```
```

Reply:
Sure I'll adjust it. In AJV there is a [`const` keyword](https://github.com/ajv-validator/ajv/blob/master/KEYWORDS.md#const) that I think will fit better here

Reply:
`const` is a great idea. I was wondering if there's something like that :)